### PR TITLE
libstd: replace all `try!` with `?` in documentation examples

### PR DESCRIPTION
**Pull request by utkarshkukreti**
See #38644.

For the record, I used the following Perl one-liner and then manually fixed a couple of things it got wrong:

    $ perl -p -i -e 's#(///.*)try!\((.*)\)#$1$2?#' src/libstd/**/*.rs

### CODE REVIEW

**Review by steveklabnik — APPROVED** [review]

File: src/libstd/io/mod.rs [hunk 1]
```diff
@@ -205,16 +205,15 @@
 //!
 //! Last, but certainly not least, is [`io::Result`]. This type is used
 //! as the return type of many `std::io` functions that can cause an error, and
-//! can be returned from your own functions as well. Many of the examples in this
-//! module use the [`try!`] macro:
+//! can be returned from your own functions as well.
 //!
```
Comment (durka) [inline src/libstd/io/mod.rs]: Why not replace <tt>Many of the examples in this module use the [`try!`] macro</tt> with <tt>Many of the examples in this module use the [`?` operator]</tt> (and link to the Syntax Index, probably)?

**Review by durka — APPROVED** [review]

File: src/libstd/io/mod.rs (hunk 1 above)
Comment (utkarshkukreti) [inline src/libstd/io/mod.rs]: I don't think that information is really relevant in this module documentation, but I'd be happy to add what you suggested if others think the same. @steveklabnik?

Comment (frewsxcv) [inline src/libstd/io/mod.rs]: I don't think that documenting `try!` or `?` is needed here, but I don't feel strongly.

**Review by frewsxcv — APPROVED** [review]

**Comment by Stebalien** [comment]
I'd suggest manually converting many of the multi-line `let a = try!(foo()); let b = try!(bar(a)); ...` blocks of code with `a()?.b()?.c()?` expressions. This was a major motivation for introducing the question mark operator.

File: src/libstd/io/mod.rs (hunk 1 above)
Comment (steveklabnik) [inline src/libstd/io/mod.rs]: My inclination is to say that this PR should be about a fairly mechanical translation, so keeping it here is fine.

Comment (utkarshkukreti) [inline src/libstd/io/mod.rs]: Done!

**Comment by utkarshkukreti** [comment]
@Stebalien I started doing that but found many cases in which I'm not sure if a single line would be clearer for the reader (I personally would write most of these in a single line in my code). For example, what do you (and others as well) think about converting these 2 cases into single lines:

```
let mut f = File::create("foo.txt")?;
f.write_all(b"Hello, world!")?;
```

-> `File::create("foo.txt")?.write_all(b"Hello, world!")?;`

and

```
let mut f = File::create("foo.txt")?;
f.set_len(10)?;
```

-> `File::create("foo.txt")?.set_len(10)?;`?

**Comment by steveklabnik** [comment]
> or example, what do you (and others as well) think about converting these 2 cases into single lines:

I tend to generally write them as two lines.

That said, I think it's best to not mess too much with the formatting of the examples here, and just worry about the transformation. So let's :shipit: 

Thank you!

@bors: r+

**Comment by josephpd3** [comment]
As someone new to the syntax, I like the decision to keep things unchained for these examples. It seems clearer visually, given the context. Looks good @utkarshkukreti!

File: src/libstd/io/mod.rs [hunk 2]
```diff
@@ -250,7 +250,7 @@
 //! [`println!`]: ../macro.println.html
 //! [`Lines`]: struct.Lines.html
 //! [`io::Result`]: type.Result.html
-//! [`try!`]: ../macro.try.html
+//! [`?` operator]: ../../book/syntax-index.html
```
Comment (pnkfelix) [inline src/libstd/io/mod.rs]: its sort of a shame that one cannot link right to the `?` operator within the index, but AFAICT there is no anchor for it specifically. (You could link to https://doc.rust-lang.org/stable/book/syntax-index.html#operators-and-symbols but I'm not sure if that pays off.)

**Comment by pnkfelix** [comment]
r=me, in case you still wanted a secondary reviewer @steveklabnik

**Comment by frewsxcv** [comment]
It looks like the bors approval didn't go through for this PR for some reason (hence it not being merged yet). @utkarshkukreti If you get some time to address the merge conflicts, we can try again. Sorry about that

**Comment by utkarshkukreti** [comment]
@frewsxcv rebased!

**Comment by frewsxcv** [comment]
Thanks!

@bors r=pnkfelix,steveklabnik,frewsxcvx

**Pull request merged**